WIP: Test cases for LayoutEditorChecks.java (code coverage )

geowar1 opened the pull request.

pabender commented:
@geowar1 there is now a conflict with this PR and master after #7639 was merged in.  I am not sure which of the two PRs should take precedence.

geowar1 commented:
This is the travis build failure… any clues?
Installing openjdk8
16810.11s$ travis_apt_get_update
169Installing APT Packages
1700.53s$ sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install graphviz libalut0 dbus-x11 tidy pulseaudio-module-x11
171Reading package lists...
172Building dependency tree...
173Reading state information...
174dbus-x11 is already the newest version (1.10.6-1ubuntu3.5).
175Some packages could not be installed. This may mean that you have
176requested an impossible situation or if you are using the unstable
177distribution that some required packages have not yet been created
178or been moved out of Incoming.
179The following information may help to resolve the situation:
180
181The following packages have unmet dependencies:
182 pulseaudio-module-x11 : Depends: libpulse0 (= 1:8.0-0ubuntu3.10) but 1:8.0-0ubuntu3.11 is to be installed
183                         Depends: pulseaudio (= 1:8.0-0ubuntu3.10)
184                         Depends: pulseaudio-utils but it is not going to be installed
185E: Unable to correct problems, you have held broken packages.
apt-get.diagnostics
186apt-get install failed
253The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install graphviz libalut0 dbus-x11 tidy pulseaudio-module-x11" failed and exited with 100 during .
254
255Your build has been stopped.

mattharris commented:
@geowar1 This seems to be a dependency problem within the Travis environment - I've raised a question over in their community for ideas:

https://travis-ci.community/t/apt-get-install-pulseaudio-module-x11-failed-in-xenial/6444

This is where they ask issues etc. to be reported.

mattharris commented:
@geowar1 right now, it seems that there is a dependency problem with Travis - restarting isn't going to help until we've understood the cause - sorry

mattharris commented:
Restarted Travis

geowar1 commented:
After all that it didn't pick up the pulseaudio fix… [/sign]

bobjacobsen commented:
Travis graphical restarted after:
```
[ERROR] Errors: 
818[ERROR]   BlockContentsIconTest.testShowIdTag:71 » TimeoutExpired Wait event queue stayi...
```
(The others completed OK)

geowar1 commented:
Did AppVeyor time out? No output after 59 minutes into the run… Why did it fail?

bobjacobsen commented:
AppVeyor has a 60 minute time limit. Most jobs take 45-49 minutes.  But several of the recent jobs have timed out after an hour:

https://ci.appveyor.com/project/JMRI/jmri/history

Did you add something that take a bunch of time?

bobjacobsen commented:
Close and reopen to set the CI branch

geowar1 commented:
AppVeyor timeout?

mattharris commented:
@geowar1 When you go to the AppVeyor build log and then login with your GitHub credentials, do you get the 'RE-BUILD PR' option?

![image](https://user-images.githubusercontent.com/738651/70894158-c40ab380-1fec-11ea-915a-1e232a33bd7e.png)

I'm trying to work out the correct set of permissions needed for this.

mattharris commented:
@geowar1 And, similarly, in Travis, when logged-in using GitHub credentials do you get the 'Restart Job' option?

![image](https://user-images.githubusercontent.com/738651/70894372-2b286800-1fed-11ea-8f2a-d5deb647fbe9.png)

rhwood commented:
@mattharris Permissions to restart jobs appears to be directly related to permission to independently commit to this project.

mattharris commented:
@rhwood Right now, you're not in any of the JMRI teams - I'll add you to JMRI\Developers to see if that changes your permissions.

rhwood commented:
@mattharris That worked.

@geowar1 You will need to explicitly logout of and back into AppVeyor and Travis CI for them to allow you to restart individual builds (I speak from experience here).

rhwood commented:
@geowar1 When dealing with test timeouts, minor Javadoc changes *will not* help reveal the issue. What does help is changing the verbosity of ant (for AppVeyor) or maven (for Travis CI) to log every test as it is run, since that will show which test started but did not stop when killed by the CI server.

For Travis CI, change line 65 of `.travis.yml` from `  - PRINT_SUMMARY=false` to `  - PRINT_SUMMARY=true`. I don't know how to make the equivalent change in ant.

Note that some JMRI developers have a strong aversion to logging anything that is not a test error or test failure, so once you figure it out, you'll need to change it back.

geowar1 commented:
> @mattharris When you go to the AppVeyor build log and then login with your GitHub credentials, do you get the 'RE-BUILD PR' option?

I do now… I'd never logged into AppVeyor; I knew about logging into the Travis to restart it. Thanks for the heads up.

geowar1 commented:
> @rhwood @geowar1 When dealing with test timeouts...

Usually when a _test_ times out its failure is logged… but when the _entire build_ times out there's just nothing in the logs after the (60-minute?) build time limit. And there's nothing in the log to tell you what ran long. So the only thing I've found to do it _try, try again_.

geowar1 commented:
@rhwood …changing the verbosity of ant (for AppVeyor) or maven (for Travis CI)…

It would be nice if that were the default for WIP builds… otherwise I'm paranoid that I'd leave that on for a production merge… (and find myself double-underlined on everyone's hate list.)

rhwood commented:
The entire build times out when a test never completes. If you want to know which test never completes, you have to turn on verbose test logging.

A typical “never completes” scenario is a modal dialog being opened and not closed, and just endlessly rerunning a build will never change its behavior.

Sent from my iPhone

> On Dec 16, 2019, at 09:53, George Warner <notifications@github.com> wrote:
> 
> ﻿
> @rhwood @geowar1 When dealing with test timeouts...
> 
> Usually when a test times out its failure is logged… but when the entire build times out there's just nothing in the logs after the (60-minute?) build time limit. And there's nothing in the log to tell you what ran long. So the only thing I've found to do it try, try again.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobjacobsen commented:
I understand that it's easier to diagnose timeouts when the last class, or even better the test, is available.  That doesn't mean that's the problem; in some cases, timeouts have been because earlier tests have started running much longer.  Ideally, logging name and duration would be ideal.  

Randall's passive-aggressive comment about "Note that some JMRI developers have a strong aversion to logging anything that is not a test error or test failure" is probably referring to me.  I have two reasons for that:

 1) I do a lot of the CI restarts of failed jobs.  I don't particularly want to make that any time-consuming than it already is.  It is _much_ harder to that in that way people have decided we should do it (putting the reason in a comment) on a phone if the log is thousands and thousands of lines long.

 2) There's a lot of value to the Unix approach to messages:  Since only warnings are emitted, they catch your eye that something has gone wrong.

If people are properly doing their own restarts, (1) goes away as a reason.  Jenkins builds using Ant and in the process counts WARN messages as errors, which would help with (2).  Jenkins also makes the full log (Log4J is quite capable of producing multiple output logs with different levels) available so perhaps something like that would help CI.

geowar1 commented:
> @bobjacobsen Note that some JMRI developers have a strong aversion to logging anything that is not a test error or test failure

Actually, I prefer less spew also… it's a lot to wade thru looking for why something failed…
How feasible would be "verbose on WIP builds"? (although that may cause Heisenbugs… ;-)

Can "testEStopButton(jmri.jmrit.throttle.ThrottleFrameTest)" be added to the intermittent tests?

geowar1 commented:
How do I resolve the conflicts? The “Resolve Conflicts” button is greyed out.

geowar1 commented:
Ok, think I figured it out… #1) GitHub desktop only has "Update from master"… so it can't "Update from <branch>" (specifically in this case "Update from development"). So I have to use command line git: "git merge development" and then edit conflicting file(s); then 'git add <file>'; then 'git commit'; and then 'git push'? Looks like it's working… (Ok, git "sucks less" today… ;-)

rhwood commented:
“Pull origin“ in GitHub Desktop and it will show every branch and origin and upstream (at least it does for me).

geowar1 commented:
I'm running GitHub desktop 2.2.4 (for macOS) and don't see "Pull origin"… is that the same as "Fetch origin"?

rhwood commented:
Likely. I wrote that from memory.

Sent from my iPhone

> On Dec 17, 2019, at 09:13, George Warner <notifications@github.com> wrote:
> 
> ﻿
> I'm running GitHub desktop 2.2.4 (for macOS) and don't see "Pull origin"… is that the same as "Fetch origin"?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobjacobsen commented:
AppVeyor timed out after an hour and restarted. Got a little further this time, which indicates that this is a “tests too long” problem, not that a single test is stuck in a loop/wait:
```
    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1951    [junit] 	at java.lang.Thread.run(Thread.java:748)
1952    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1953    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1954    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1955    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1956    [junit] java.lang.NullPointerException
1957    [junit] 	at sun.awt.windows.WWindowPeer.getActiveWindowHandles(WWindowPeer.java:463)
1958    [junit] 	at sun.awt.windows.WToolkit.eventLoop(Native Method)
1959    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1960    [junit] 	at java.lang.Thread.run(Thread.java:748)
1961    [junit] INFO  - This INFO message was emitted to test the entire logging chain, please don't remove [main] jmri.util.JUnitAppenderTest.testUnexpectedCheck()
1962    [junit] java.lang.NullPointerException
1963    [junit] 	at sun.awt.windows.WWindowPeer.getActiveWindowHandles(WWindowPeer.java:463)
1964    [junit] 	at sun.awt.windows.WToolkit.eventLoop(Native Method)
1965    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1966    [junit] 	at java.lang.Thread.run(Thread.java:748)
```

bobjacobsen commented:
AppVeyor timed out. Last messages:
```
    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1950    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1951    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1952    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1953    [junit] java.lang.NullPointerException
1954    [junit] 	at sun.awt.windows.WWindowPeer.getActiveWindowHandles(WWindowPeer.java:463)
1955    [junit] 	at sun.awt.windows.WToolkit.eventLoop(Native Method)
1956    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1957    [junit] 	at java.lang.Thread.run(Thread.java:748)
1958    [junit] INFO  - This INFO message was emitted to test the entire logging chain, please don't remove [main] jmri.util.JUnitAppenderTest.testUnexpectedCheck()
1959    [junit] java.lang.NullPointerException
1960    [junit] 	at sun.awt.windows.WWindowPeer.getActiveWindowHandles(WWindowPeer.java:463)
1961    [junit] 	at sun.awt.windows.WToolkit.eventLoop(Native Method)
1962    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1963    [junit] 	at java.lang.Thread.run(Thread.java:748)
```

bobjacobsen commented:
Resolved a conflict so tests will restart. @geowar1, you'll need to pull this branch back from GitHub to pick up that change.

geowar1 commented:
I'm now getting this error:
java/test/jmri/jmrit/display/layoutEditor/LayoutEditorTest.java:51: error: reference to Test is ambiguous
    @Test
  both class org.netbeans.jemmy.Test in org.netbeans.jemmy and class org.junit.Test in org.junit 

I removed the wild card imports ("import import org.netbeans.jemmy.*" and "import org.junit.*") and added the implicit: ("import org.junit.Test;") and I'm still seeing the error.
Little help?

geowar1 commented:
Nevermind... figured it out: used both the wild card and the implicit:
import org.junit.*;
import org.junit.Test;

geowar1 commented:
Don’t know why that stopped working… none of the other test files had to use the implicit import.

geowar1 commented:
AppVeyor:jmri.jmrit.throttle.ThrottleFrameTest.testStopButton() failed. Restarting…

bobjacobsen commented:
This is also showing conflicts in java/test/jmri/jmrit/display/layoutEditor/LayoutEditorToolsTest.java that seem related to the ones in #7730

bobjacobsen commented:
AppVeyor failed without end report.  Close/Reopen to pick up @pabender's fix of that.

bobjacobsen commented:
Is this ready to merge?  I'd like to get it merged as soon as CI is passed. If not, could you mark it WIP please?

geowar1 commented:
Ready to merge… just waiting for all checks to pass.

bobjacobsen commented:
@pabender - There are several PRs that are consistently failing AppVeyor, but not Travis.  They're not providing the end-of-log summary.  They've been closed and reopened to pick up your recent fix.  An example log is here:  https://ci.appveyor.com/project/JMRI/jmri/builds/29709157

The end of one of the logs:
```
1962    [junit] INFO  - No pre-existing config file found, searched for 'C:\projects\jmri\temp\DecoderProConfig3.xml' [main] apps.AppsBase.setAndLoadPreferenceFile()
1963    [junit] INFO  - No pre-existing config file found, searched for 'C:\projects\jmri\temp\DecoderProConfig3.xml' [main] apps.AppsBase.setAndLoadPreferenceFile()
1964    [junit] Tests run: 25240, Failures: 0, Errors: 3, Skipped: 827, Time elapsed: 1,845.516 sec
1965mvn :     [junit] Test apps.tests.AllTest FAILED
1966At line:1 char:1
1967+ mvn --quiet test -P appveyor --batch-mode
1968+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1969    + CategoryInfo          : NotSpecified: (    [junit] Tes....AllTest FAILED:String) [], RemoteException
1970    + FullyQualifiedErrorId : NativeCommandError
1971 
1972[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.8:run (test) on project jmri: An Ant BuildException has occured: The following error occurred while executing this line:
1973[ERROR] C:\projects\jmri\build.xml:1556: condition satisfied
1974[ERROR] around Ant part ...<ant antfile="C:\projects\jmri/build.xml">... @ 4:45 in C:\projects\jmri\target\antrun\build-main.xml
1975[ERROR] -> [Help 1]
1976[ERROR] 
1977[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
1978[ERROR] Re-run Maven using the -X switch to enable full debug logging.
1979[ERROR] 
1980[ERROR] For more information about the errors and possible solutions, please read the following articles:
1981[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
1982Command executed with exception:     [junit] Test apps.tests.AllTest FAILED
1983appveyor PushArtifact tests.log
1984Uploading artifact tests.log (374,421 bytes)...100%
1985appveyor PushArtifact junit-results.xml
1986File not found: C:\projects\jmri\junit-results.xml
1987Command exited with code 2
```

Any ideas on how to make progress on this?

rhwood commented:
I’m not 100% on AppVeyor (re)pulling from new head of branch-to-merge-into in the way Travis CI does when closing and reopening a PR. I think that if there is a fix in development that needs to be pulled into, its more reliable to merge development into the PR branch.

Randall
Sent from my iPhone

> On Dec 22, 2019, at 17:59, Bob Jacobsen <notifications@github.com> wrote:
> 
> ﻿
> @pabender - There are several PRs that are consistently failing AppVeyor, but not Travis. They're not providing the end-of-log summary. They've been closed and reopened to pick up your recent fix. An example log is here: https://ci.appveyor.com/project/JMRI/jmri/builds/29709157
> 
> The end of one of the logs:
> 
> 1962    [junit] INFO  - No pre-existing config file found, searched for 'C:\projects\jmri\temp\DecoderProConfig3.xml' [main] apps.AppsBase.setAndLoadPreferenceFile()
> 1963    [junit] INFO  - No pre-existing config file found, searched for 'C:\projects\jmri\temp\DecoderProConfig3.xml' [main] apps.AppsBase.setAndLoadPreferenceFile()
> 1964    [junit] Tests run: 25240, Failures: 0, Errors: 3, Skipped: 827, Time elapsed: 1,845.516 sec
> 1965mvn :     [junit] Test apps.tests.AllTest FAILED
> 1966At line:1 char:1
> 1967+ mvn --quiet test -P appveyor --batch-mode
> 1968+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1969    + CategoryInfo          : NotSpecified: (    [junit] Tes....AllTest FAILED:String) [], RemoteException
> 1970    + FullyQualifiedErrorId : NativeCommandError
> 1971 
> 1972[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.8:run (test) on project jmri: An Ant BuildException has occured: The following error occurred while executing this line:
> 1973[ERROR] C:\projects\jmri\build.xml:1556: condition satisfied
> 1974[ERROR] around Ant part ...<ant antfile="C:\projects\jmri/build.xml">... @ 4:45 in C:\projects\jmri\target\antrun\build-main.xml
> 1975[ERROR] -> [Help 1]
> 1976[ERROR] 
> 1977[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> 1978[ERROR] Re-run Maven using the -X switch to enable full debug logging.
> 1979[ERROR] 
> 1980[ERROR] For more information about the errors and possible solutions, please read the following articles:
> 1981[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> 1982Command executed with exception:     [junit] Test apps.tests.AllTest FAILED
> 1983appveyor PushArtifact tests.log
> 1984Uploading artifact tests.log (374,421 bytes)...100%
> 1985appveyor PushArtifact junit-results.xml
> 1986File not found: C:\projects\jmri\junit-results.xml
> 1987Command exited with code 2
> Any ideas on how to make progress on this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobjacobsen commented:
Thanks.  I've pulled `development` into #7755 to see if that helps.  But it could be that AppVeyor is just broken

> I’m not 100% on AppVeyor (re)pulling from new head of branch-to-merge-into in the way Travis CI does when closing and reopening a PR. I think that if there is a fix in development that needs to be pulled into, its more reliable to merge development into the PR branch.

rhwood commented:
We should pay close attention; AppVeyor recently changed their default image to Windows Server 2019 (the server version of Windows 10 1809). It is possible that change is causing issues.

We could also be running into a Windows specific failure of some recent changes in JMRI.

I'll see if I can spot something in the noise tomorrow; it would also be helpful if someone with access to Windows 10 could run `ant alltest`.

Randall
Sent from my iPad

> On Dec 22, 2019, at 8:49 PM, Bob Jacobsen <notifications@github.com> wrote:
> 
> Thanks. I've pulled development into #7755 to see if that helps. But it could be that AppVeyor is just broken
> 
> I’m not 100% on AppVeyor (re)pulling from new head of branch-to-merge-into in the way Travis CI does when closing and reopening a PR. I think that if there is a fix in development that needs to be pulled into, its more reliable to merge development into the PR branch.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobjacobsen commented:
I ran a modified AppVeyor job to do `ant alltest` against the head of `development`.  The log is attached here.
[log-2.txt](https://github.com/JMRI/JMRI/files/3993664/log-2.txt)
 It showed multiple errors, see below.  This might indicate that there's something wrong at the HEAD of `development`.

```
2087     [java] There were 6 failures:
2088     [java] 1) testToolBarPositionBottom(jmri.jmrit.display.layoutEditor.LayoutEditorTest)
2089     [java] org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
2090     [java] 	at java.lang.Thread.sleep(Native Method)
2091     [java] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:206)
2092     [java] 	at org.netbeans.jemmy.ActionProducer.produceAction(ActionProducer.java:265)
2093     [java] 	at org.netbeans.jemmy.operators.Operator.produceTimeRestricted(Operator.java:769)
2094     [java] 	at org.netbeans.jemmy.operators.Operator.produceTimeRestricted(Operator.java:794)
2095     [java] 	at org.netbeans.jemmy.operators.JMenuOperator.pushMenu(JMenuOperator.java:333)
2096     [java] 	at org.netbeans.jemmy.operators.JMenuOperator.pushMenu(JMenuOperator.java:383)
2097     [java] 	at org.netbeans.jemmy.operators.JMenuOperator.pushMenu(JMenuOperator.java:436)
2098     [java] 	at org.netbeans.jemmy.operators.JMenuOperator.pushMenu(JMenuOperator.java:552)
2099     [java] 	at jmri.jmrit.display.layoutEditor.LayoutEditorTest.testToolBarPositionBottom(LayoutEditorTest.java:938)
2100     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2101     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2102     [java] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2103     [java] 	at java.lang.reflect.Method.invoke(Method.java:498)
2104     [java] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2105     [java] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2106     [java] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2107     [java] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2108     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2109     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2110     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2111     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2112     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2113     [java] 	at java.lang.Thread.run(Thread.java:748)
2114     [java] 2) testSetSignalHeadOnPanelAtXYDoubleAndRemove(jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest)
2115     [java] org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
2116     [java] 	at java.lang.Thread.sleep(Native Method)
2117     [java] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:206)
2118     [java] 	at org.netbeans.jemmy.QueueTool.waitEmpty(QueueTool.java:307)
2119     [java] 	at jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest.testSetSignalHeadOnPanelAtXYDoubleAndRemove(LayoutEditorToolsTest.java:530)
2120     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2121     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2122     [java] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2123     [java] 	at java.lang.reflect.Method.invoke(Method.java:498)
2124     [java] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2125     [java] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2126     [java] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2127     [java] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2128     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2129     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2130     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2131     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2132     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2133     [java] 	at java.lang.Thread.run(Thread.java:748)
2134     [java] 3) testSetSignalHeadOnPanelAtXYIntAndRemove(jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest)
2135     [java] org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
2136     [java] 	at java.lang.Thread.sleep(Native Method)
2137     [java] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:206)
2138     [java] 	at org.netbeans.jemmy.QueueTool.waitEmpty(QueueTool.java:307)
2139     [java] 	at jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest.testSetSignalHeadOnPanelAtXYIntAndRemove(LayoutEditorToolsTest.java:493)
2140     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2141     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2142     [java] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2143     [java] 	at java.lang.reflect.Method.invoke(Method.java:498)
2144     [java] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2145     [java] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2146     [java] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2147     [java] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2148     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2149     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2150     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2151     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2152     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2153     [java] 	at java.lang.Thread.run(Thread.java:748)
2154     [java] 4) testSetSignalHeadOnPanelAtPointAndRemove(jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest)
2155     [java] org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
2156     [java] 	at java.lang.Thread.sleep(Native Method)
2157     [java] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:206)
2158     [java] 	at org.netbeans.jemmy.QueueTool.waitEmpty(QueueTool.java:307)
2159     [java] 	at jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest.testSetSignalHeadOnPanelAtPointAndRemove(LayoutEditorToolsTest.java:512)
2160     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2161     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2162     [java] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2163     [java] 	at java.lang.reflect.Method.invoke(Method.java:498)
2164     [java] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2165     [java] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2166     [java] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2167     [java] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2168     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2169     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2170     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2171     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2172     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2173     [java] 	at java.lang.Thread.run(Thread.java:748)
2174     [java] 5) testWarrantLoopRun(jmri.jmrit.logix.NXFrameTest)
2175     [java] org.junit.runners.model.TestTimedOutException: test timed out after 20 seconds
2176     [java] 	at java.lang.Thread.sleep(Native Method)
2177     [java] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:206)
2178     [java] 	at org.netbeans.jemmy.QueueTool.waitEmpty(QueueTool.java:307)
2179     [java] 	at jmri.jmrit.logix.NXFrameTest.runtimes(NXFrameTest.java:384)
2180     [java] 	at jmri.jmrit.logix.NXFrameTest.testWarrantLoopRun(NXFrameTest.java:294)
2181     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2182     [java] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2183     [java] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2184     [java] 	at java.lang.reflect.Method.invoke(Method.java:498)
2185     [java] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2186     [java] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2187     [java] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2188     [java] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2189     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2190     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2191     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2192     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2193     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2194     [java] 	at java.lang.Thread.run(Thread.java:748)
2195     [java] 6) testNXWarrant(jmri.jmrit.logix.NXFrameTest)
2196     [java] org.junit.runners.model.TestTimedOutException: test timed out after 20 seconds
2197     [java] 	at sun.misc.Unsafe.park(Native Method)
2198     [java] 	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
2199     [java] 	at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:426)
2200     [java] 	at java.util.concurrent.FutureTask.get(FutureTask.java:204)
2201     [java] 	at org.junit.internal.runners.statements.FailOnTimeout.getResult(FailOnTimeout.java:141)
2202     [java] 	at org.junit.internal.runners.statements.FailOnTimeout.evaluate(FailOnTimeout.java:127)
2203     [java] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2204     [java] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2205     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2206     [java] 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2207     [java] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2208     [java] 	at java.lang.Thread.run(Thread.java:748)
2209     [java] 
2210     [java] FAILURES!!!
2211     [java] Tests run: 24984,  Failures: 6
```

bobjacobsen commented:
Running on Mac Mojave via `./runtest.csh jmri.jmrit.display.layoutEditor.LayoutEditorTest`, the `testToolBarPositionBottom(jmri.jmrit.display.layoutEditor.LayoutEditorTest)` failure repeats consistently.

Running `./runtest.csh jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest`, I consistently get a different failure:
```
There was 1 failure:
1) testSetSignalsAtTurnoutWithDone(jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest)
org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.netbeans.jemmy.QueueTool.invokeAndWait(QueueTool.java:436)
	at org.netbeans.jemmy.QueueTool.invokeSmoothly(QueueTool.java:372)
	at org.netbeans.jemmy.operators.Operator.runMappingPrimitive(Operator.java:1003)
	at org.netbeans.jemmy.operators.Operator.runMapping(Operator.java:958)
	at org.netbeans.jemmy.operators.AbstractButtonOperator.doClick(AbstractButtonOperator.java:532)
	at jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest.testSetupSSL(LayoutEditorToolsTest.java:303)
	at jmri.jmrit.display.layoutEditor.LayoutEditorToolsTest.testSetSignalsAtTurnoutWithDone(LayoutEditorToolsTest.java:222)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

bobjacobsen commented:
PR #7761 is an attempt to fix this by setting `@Ignore` on the failing tests.

bobjacobsen commented:
`Try RetryRule at 5` isn't a winning strategy:  If three retries aren't enough, there really is something wrong.  #7761 is fixing some permanently-failing tests in the current head of `development`, so perhaps you're seeing those, or perhaps new ones.

bobjacobsen commented:
Resolved minor conflict uncovered by #7761 

If this passes CI, is it ready to merge?

bobjacobsen commented:
Close/reopen to ensure picks up #7761

bobjacobsen commented:
Marking this as WIP so the other PRs get merged first. Please take that off the title line once this is set to go. Thanks!

geowar1 commented:
[junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490

geowar1 commented:
[WARNING] Rule violated for source file jmri/jmrit/display/layoutEditor/LayoutEditorVerticalToolBarPanel.java: lines covered ratio is 0.00000, but expected minimum is 0.00001

Fixing via PR #7770

bobjacobsen commented:
Close and reopen to pick up #7770

geowar1 commented:
jmri.jmrit.operations.setup.OperationsSetupFrameTest          testSetupFrameWrite <error message="Wait AbstractButton with text "HO" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.TimeoutEx...
1790jmri.jmrit.operations.automation.actions.BuildTrainActionTest testActionMessages  <error message="Wait AbstractButton with text "Halt" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.Timeout...
1791jmri.jmrit.timetable.swing.TimeTableFrameTest                 testDriver          <error message="Wait "Item "Station 2" available in combo box." state to be reached (ComponentOperator.WaitStateTimeout)" ty...

pabender commented on java/src/jmri/jmrit/display/layoutEditor/LayoutTurnout.java:
```diff
@@ -2003,7 +2001,7 @@ protected int findHitPointType(@Nonnull Point2D hitPoint, boolean useRectangles,
 
         double circleRadius = LayoutEditor.SIZE * layoutEditor.getTurnoutCircleSize();
-        double distance, minDistance = POSITIVE_INFINITY;
+        double distance, minDistance = Float.POSITIVE_INFINITY;
```
There is a type mismatch here.  The value is a double, but you are assigning a Float value.  You should replace Float.POSITIVE_INFINITY with Double. POSITIVE_INFINITY

geowar1 replied:
Thanks for the catch… fixed.

pabender commented on java/src/jmri/jmrit/display/layoutEditor/TrackSegment.java:
```diff
@@ -644,7 +639,7 @@ protected int findHitPointType(Point2D hitPoint, boolean useRectangles, boolean
 
             double circleRadius = LayoutEditor.SIZE * layoutEditor.getTurnoutCircleSize();
-            double distance, minDistance = POSITIVE_INFINITY;
+            double distance, minDistance = Float.POSITIVE_INFINITY;
```
There is a type mismatch here. The value is a double, but you are assigning a Float value. You should replace Float.POSITIVE_INFINITY with Double. POSITIVE_INFINITY

pabender replied:
You did not fix this one, you just marked the conversation resolved

geowar1 replied:
Sorry, missed that one… next commit should fix it.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,349 @@
+    public Timeout globalTimeout = Timeout.seconds(10);
+
+    @Rule   // allow 5 retries
+    public RetryRule retryRule = new RetryRule(5);
```
I think 5 is way too high.  If a test needs 5 retries it is really too flakey to be reliable.  I really don’t want to see this higher than 3.

geowar1 replied:
Agreed… fixed.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,349 @@
-    public void setUp() {
+    @Test
+    public void testSetup() {
+        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
```
I don’t think this test that the setup method completed belongs here. If the setup fails, then tests that use the configured components will fail.  That is the expected behavior.

geowar1 replied:
I can live with that… Fixed.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,349 @@
+    }
+
+    @Test
+    public void testBundleStrings() {
```
Testing package Specific bundle strings really should be in the bundle tests, not here.

geowar1 replied:
Personally, I'd prefer to test the ones used in the test class that uses them… but I'm not married to it so… removed.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,349 @@
+    }   //testCheckUnnecessaryAnchors
+
+    @BeforeClass
+    public static void setUpClass() {
```
I strongly feel that this setup should be done in a per-test setup method rather than in a static method.

It is too easy to have one test contaminate the environment of a following test if they are all using the same configuration.

geowar1 replied:
The items setup here should not be effected by any tests… if they are then that's a failure and I want to catch it. Doing this before every test would hide that error and also just waste time. (The LayoutEditorChecks shouldn't effect anything: they only report possible errors.)

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,349 @@
+            JUnitUtil.resetProfileManager();
+
+            // set default string matching comparator to one that exactly matches and is case sensitive
+            Operator.setDefaultStringComparator(new Operator.DefaultStringComparator(true, true));
```
It is entirely possible that this operation on the operator has global implications.  This actually MIGHT be causing failures in tests outside this test class.

Make sure you reset this to the previous default when cleanup occurs.

geowar1 replied:
That was my thought as well… (long after the fact). I've added the restore code to most of my PR's that change it (excluding the one that you caught). (IMHO: true, true should be the default… why would you want sloppy matching?!? My initial work-around (before I found this setting) was I had to hard-code indexes into possible matches. Setting this true, true eliminated the hard coded indexes… which IMHO: is a win.)

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorDialogs/EnterReporterDialogTest.java:
```diff
@@ -38,6 +35,9 @@
         JUnitUtil.setUp();
         if (!GraphicsEnvironment.isHeadless()) {
+            // set default string matching comparator to one that exactly matches and is case sensitive
+            Operator.setDefaultStringComparator(new Operator.DefaultStringComparator(true, true));
```
It is entirely possible that this operation on the operator has global implications. This actually MIGHT be causing failures in tests outside this test class.

Make sure you reset this to the previous default when cleanup occurs.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorDialogs/LayoutTrackEditorsTest.java:
```diff
@@ -815,12 +882,15 @@ public void createBlocks() {
+            InstanceManager.setDefault(BlockManager.class, new BlockManager());
+
+            // set default string matching comparator to one that exactly matches and is case sensitive
+            Operator.setDefaultStringComparator(new Operator.DefaultStringComparator(true, true));
```
It is entirely possible that this operation on the operator has global implications. This actually MIGHT be causing failures in tests outside this test class.

Make sure you reset this to the previous default when cleanup occurs.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorTest.java:
```diff
@@ -23,11 +22,11 @@
-    @Rule
-    public RetryRule retryRule = new RetryRule(3); // allow 3 retries
+    @Rule   // allow 5 retries
+    public RetryRule retryRule = new RetryRule(5);
```
5 is too high.

geowar1 replied:
Fixed.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorToolsTest.java:
```diff
@@ -597,10 +685,14 @@ public void setUp() throws Exception {
             JUnitUtil.resetProfileManager();
 
+            // set default string matching comparator to one that exactly matches and is case sensitive
+            Operator.setDefaultStringComparator(new Operator.DefaultStringComparator(true, true));
```
It is entirely possible that this operation on the operator has global implications. This actually MIGHT be causing failures in tests outside this test class.

Make sure you reset this to the previous default when cleanup occurs.

pabender commented on java/src/jmri/jmrit/display/layoutEditor/LayoutEditorChecks.java:
```diff
@@ -346,7 +350,9 @@ private void setupCheckNonContiguousBlocksMenu() {
         HashMap<String, List<Set<String>>> blockNamesToTrackNameSetMaps = new HashMap<>();
         for (LayoutTrack layoutTrack : layoutEditor.getLayoutTracks()) {
-            layoutTrack.checkForNonContiguousBlocks(blockNamesToTrackNameSetMaps);
+            if (layoutTrack != null) {
```
I don’t think you need the null check here.  You are iterating over a collection of layoutTrack objects

geowar1 replied:
One of the LayoutTrack tests was crashing here with a NPE… I don't know how the null got in the list so I put in the check. If(/when) I get appveyor passing again I can spend time rooting out where the null came from… I won't close this as resolved.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorDialogs/LayoutTrackEditorsTest.java:
```diff
@@ -48,6 +50,13 @@ public void testCTor() {
     }
 
+    @Test
+    public void testBundleStrings() {
```
You removed this from one file and added it here too.  It should also be in the bundle test.

geowar1 replied:
Removing…

geowar1 commented:
>Please move the unrelated throttle changes to a separate pull request. They don’t belong in this one.

The throttle changes have been removed; original file restore (but imports fixed).
